File: src/game/pool.ts
// Builds the set of eligible quiz countries from the difficulty filters.
// Only countries that exist on the map AND have a flag/metadata are eligible.
import { countries, sizeBucket, type CountryMeta } from '../data/countries';
import { mapIds } from '../map/world';
import type { DifficultyFilter } from '../store/settingsStore';

export function buildPool(filter: DifficultyFilter): CountryMeta[] {
  return countries.filter((c) => {
    if (!mapIds.has(c.id)) return false;
    if (filter.continents.length && !filter.continents.includes(c.continent)) {
      return false;
    }
    if (filter.size !== 'all' && sizeBucket(c.area) !== filter.size) {
      return false;
    }
    return true;
  });
}

function shuffle<T>(arr: T[]): T[] {
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr;
}

// Shuffle-bag picker. Pure random-with-replacement clusters badly (some
// countries appear 5× while others never show), which feels broken. Instead we
// shuffle the whole pool and draw without replacement, so every country comes up
// once before any repeat. The bag is rebuilt when the pool changes (key) or runs
// out, and we avoid an immediate repeat across the refill boundary.
let bag: CountryMeta[] = [];
let bagKey = '';

export function nextFromBag(
  pool: CountryMeta[],
  key: string,
  lastId?: string,
): CountryMeta | null {
  if (pool.length === 0) return null;
  if (pool.length === 1) return pool[0];

  if (key !== bagKey || bag.length === 0) {
    bag = shuffle(pool.slice());
    bagKey = key;
    // The next draw pops from the end; make sure it isn't the country we just
    // showed (can happen right after a refill).
    if (lastId && bag[bag.length - 1].id === lastId) {
      [bag[bag.length - 1], bag[0]] = [bag[0], bag[bag.length - 1]];
    }
  }
  return bag.pop() ?? null;
}

File: server/src/game/pool.rs
//! Builds the eligible country pool from a difficulty filter and produces the
//! flag sequence for a match. The server owns the sequence so every client in a
//! room races the identical flags. Mirrors src/game/pool.rs.
use rand::seq::SliceRandom;
use rand::Rng;

use super::{map_ids, Country, COUNTRIES};
use crate::protocol::DifficultyFilter;

/// Area thresholds (km²) for the "by size" filter — must match
/// src/data/countries.ts SIZE_THRESHOLDS.
const SMALL_MAX: f64 = 50_000.0;
const LARGE_MIN: f64 = 1_000_000.0;

fn size_bucket(area: f64) -> &'static str {
    if area < SMALL_MAX {
        "small"
    } else if area > LARGE_MIN {
        "large"
    } else {
        "medium"
    }
}

/// Countries eligible under the filter: on the map, in an allowed continent and
/// size bucket, and within the recognition scope. Mirrors `buildPool`.
pub fn build_pool(filter: &DifficultyFilter) -> Vec<&'static Country> {
    let ids = map_ids();
    let scope_un = filter.scope.as_deref() == Some("un");
    COUNTRIES
        .iter()
        .filter(|c| {
            if !ids.contains(c.id) {
                return false;
            }
            if !filter.continents.is_empty() && !filter.continents.iter().any(|x| x == c.continent)
            {
                return false;
            }
            if filter.size != "all" && size_bucket(c.area) != filter.size {
                return false;
            }
            if scope_un && !c.un_member {
                return false;
            }
            true
        })
        .collect()
}

/// Produce `n` targets using a shuffle bag: shuffle the pool and draw without
/// replacement (refilling when exhausted) so every country appears once before
/// any repeat, and never twice in a row across a refill. Mirrors `nextFromBag`.
pub fn make_sequence<R: Rng>(pool: &[&'static Country], n: usize, rng: &mut R) -> Vec<&'static Country> {
    let mut seq: Vec<&'static Country> = Vec::with_capacity(n);
    if pool.is_empty() {
        return seq;
    }
    if pool.len() == 1 {
        return vec![pool[0]; n];
    }
    let mut bag: Vec<&'static Country> = Vec::new();
    while seq.len() < n {
        if bag.is_empty() {
            bag = pool.to_vec();
            bag.shuffle(rng);
            // Avoid an immediate repeat across the refill boundary.
            if let (Some(last), Some(&next)) = (seq.last(), bag.last()) {
                if last.id == next.id {
                    let len = bag.len();
                    bag.swap(len - 1, 0);
                }
            }
        }
        seq.push(bag.pop().unwrap());
    }
    seq
}
